### ISSUE TITLE
`UNKNOWN ERROR Original error: 'image'` when rerunning a failed Cellpose task

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#842]
I'm hitting the following error:

```
UNKNOWN ERROR
Original error: 'image'
```

Steps to reproduce:
Have a workflow that runs into a job execution error (e.g. cellpose runs out of memory)
Rerun only the cellpose task using the existing output dataset
Immediately hits that error

**Comment by tcompa** [comment]
Possibly a bug introduced with #803, in the history.

Any chance you can look at the output-dataset metadata after the unkown-error failure? My suspect is that the `image` key is missing.

**Comment by jluethi** [comment]
But we do get this workflow log and it contains the history. Can you check based on this one?

[workflow.log](https://github.com/fractal-analytics-platform/fractal-server/files/12606910/workflow.log)

**Comment by jluethi** [comment]
I think your suspicion may be on to something here. After rerunning, the dataset meta only contains history, see here:

<img width="1563" alt="Screenshot 2023-09-14 at 11 07 56" src="https://github.com/fractal-analytics-platform/fractal-server/assets/18033446/dd72e2e5-3700-4c43-86e4-cc3b5176459b">

It usually has many more entries above history, like Plate, Well, Image etc.

**Comment by jluethi** [comment]
Ah, this may be relevant:
The error I hit in cellpose was an out of memory, but it hit a TaskError:
```
TASK ERROR:Task id: 246 (Cellpose Segmentation), e.workflow_task_order=8
TRACEBACK:
Traceback (most recent call last):
  File "/data/homes/fractal/anaconda3/envs/fractal-server-1.3.5/lib/python3.9/site-packages/fractal_server/app/runner/_common.py", line 391, in call_single_parallel_task
    raise e
  File "/data/homes/fractal/anaconda3/envs/fractal-server-1.3.5/lib/python3.9/site-packages/fractal_server/app/runner/_common.py", line 384, in call_single_parallel_task
    _call_command_wrapper(
  File "/data/homes/fractal/anaconda3/envs/fractal-server-1.3.5/lib/python3.9/site-packages/fractal_server/app/runner/_common.py", line 183, in _call_command_wrapper
    raise TaskExecutionError(err)

lots of cellpose logging

Traceback (most recent call last):
  File "/net/nfs4/pelkmanslab-fileserver-fractal/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/cellpose_segmentation.py", line 691, in <module>
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/_utils.py", line 79, in run_fractal_task
  File "pydantic/decorator.py", line 40, in pydantic.decorator.validate_arguments.validate.wrapper_function
  File "pydantic/decorator.py", line 134, in pydantic.decorator.ValidatedFunction.call
  File "pydantic/decorator.py", line 206, in pydantic.decorator.ValidatedFunction.execute
  File "/net/nfs4/pelkmanslab-fileserver-fractal/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/cellpose_segmentation.py", line 582, in cellpose_segmentation
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/fractal_tasks_core/lib_masked_loading.py", line 239, in masked_loading_wrapper
  File "/net/nfs4/pelkmanslab-fileserver-fractal/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/cellpose_segmentation.py", line 113, in segment_ROI
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/cellpose/models.py", line 556, in eval
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/cellpose/models.py", line 645, in _run_cp
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/cellpose/dynamics.py", line 743, in compute_masks
  File "/data/homes/fractal/20230627_joel_fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.11.0a2/venv/lib/python3.9/site-packages/cellpose/dynamics.py", line 690, in get_masks
numpy.core._exceptions.MemoryError: Unable to allocate 864. KiB for an array with shape (110592,) and data type int64
```

**Comment by jluethi** [comment]
It appears that this numpy memory error triggers some unexpected behavior downstream. So there may be special types of TaskErrors that are the issue here

**Comment by tcompa** [comment]
Thanks @jluethi. I do observe something which looks wrong in #803, but it's slightly at odd with your reporting.
It seems that _any_ error (e.g. both TaskExecutionError and JobExecutionError) is not setting `output_dataset.meta` correctly - and not just the JobExecutionError. If this is true, then the different behaviors you observe are due to different initial conditions (e.g. whether the `image` key is or is not present in the DB).

I may also be wrong.. I'm digging further into it now.

**Comment by jluethi** [comment]
I think it's only for TaskExecutionError where somehow the whole meta gets overwritten / not correctly written and we only get history in the meta. For JobExecutionError, it actually works as expected (sorry for the confusion above)

I can make a workflow fail in some scenarios and it still has the correct meta set afterwards, even though the job failed. For example, if I run the 01 example and set the mem for task 2 way too low, then task 2 fails. But I can just rerun task 2 (as would be the goal). That is the scenario of a JobExecutionError.

I also tried to produce a trivial TaskExecutionError by providing wrong ROI tables in the copy task. That then lead to meta being set wrong (only history). Thus, rerunning that job the fails again (the error is slightly different, because it was a job that was not parallelized, but the root cause is the same I'd think)

**Comment by jluethi** [comment]
I could create a minimal example that fully reproduces the bug above:
Trigger a trivial TaskError: In example 01, set the level to an invalid target (e.g. 7, when only 5 pyramid levels exist).

This leads to a TaskError.

Trying to rerun that cellpose task with correct pyramid levels based on the output dataset then fails with the UNKNOWN ERROR Original error: 'image' from above

**Comment by tcompa** [comment]
I think this boils down to this old discussion:

> If a workflow fails, should the metadata be updated?
> What we can do now: Update to the last valid state, i.e. to the state at the end of step 2.
> Writing the last valid state
> (https://github.com/fractal-analytics-platform/fractal-server/issues/261#issuecomment-1559130894)

which (as far as I remember) was never implemented.

The relevant _open_ issue is https://github.com/fractal-analytics-platform/fractal-server/issues/788, which says
> CURRENTLY DISCARDED OPTION (*): When a workflow fails with a TaskExecutionError happens, update output_dataset.meta (in the DB)
> ...
> (\*) The idea would be that if execution failed, the only way to continue is via the specific "resume" feature, and not via the standard workflow execution. This can be discussed (see also https://github.com/fractal-analytics-platform/fractal-server/issues/261#issuecomment-1559130894).

**Comment by tcompa** [comment]
Current status:
1. This seems unrelated to #803, meaning that the "a failed job erases the `Dataset.meta["image"]` key/value pair" guess seems wrong.
2. Issue #788 (resuming execution after a failed task) was never implemented; now that the A-to-B feature is being tested a bit more, we realized that some scenarios are not covered. My (non-verified) understanding is that:
    * If you start with a dataset which already has the `image` component list in the metadata (e.g. because you already successfully ran the parsing tasks sub-workflow), then the issue described here often won't matter (the "often" is because there may be caveats related to copy-ome-zarr);
    * If you run a whole workflow (image-parsing included) and it fails at the cellpose task, you won't be able to restart execution

The original formulation of #788 has a broad scope, and it was also re-discussed later ("Let's put this on hold, since we had multiple ideas on how to proceed (partly to be implemented in fractal-server, and partly in fractal-web)."). Therefore it is not ready to be implemented without a thorough review.

The easiest workaround is the one described in #844.

**Comment by tcompa** [comment]
This should be fixed in `1.3.6a0`, but it's valuable to test the new version in a few realistic scenarios.

**Comment by jluethi** [comment]
Confirming this as fixed in 1.6.0a0! :) Tested on UZH deployment with different submission scenarios, can't reproduce the issue anymore and found no new issues :)

**Comment by tcompa** [comment]
Now released as 1.3.6. Closing.
